### ISSUE TITLE
Error: Connection drops, seemingly after making a first connection.

### ISSUE THREAD

**Issue by Texlo-Dev** [benashford/redis-async-rs#32]
Hi Ben,
I've attempted to use your library to make a pair of redis connections for my service. The first connection is successful, but the second one fails. 
After logging the errors: I am greeted with the following:
```
[2019-02-19T00:53:30Z DEBUG event_worker::handler] Beginning connection to Redis.
[2019-02-19T00:53:30Z INFO  redis_async::reconnect] Connection established
[2019-02-19T00:53:30Z DEBUG event_worker::handler] Making second connection to Redis.
[2019-02-19T00:53:30Z INFO  redis_async::reconnect] Connection established
[2019-02-19T00:53:30Z ERROR redis_async::client::paired] Connection to Redis closed unexpectedly
[2019-02-19T00:53:30Z DEBUG tokio_reactor] dropping I/O source: 0
[2019-02-19T00:53:30Z WARN  event_worker::handler] There was an error while attempting to receive a redis event. Redis(Remote("ERR Protocol error: expected \'$\', got \':\'"))
[2019-02-19T00:53:30Z ERROR redis_async::reconnect] Cannot perform action: Cannot write to channel: send failed because receiver is gone
[2019-02-19T00:53:30Z WARN  event_worker::handler] There was an error while attempting to receive a redis event. Redis(EndOfStream)
[2019-02-19T00:53:30Z WARN  event_worker::handler] There was an error while attempting to receive a redis event. Redis(EndOfStream)
[2019-02-19T00:53:30Z WARN  event_worker::handler] There was an error while attempting to receive a redis event. Redis(EndOfStream)
```

**My Code:**  https://gist.github.com/Texlo-Dev/011a5f34556ca50ff7fa4675b56f9446
Any help is much appreciated.

Thanks,
Texlo-Dev

**Comment by benashford** [comment]
Hello,

Thanks for the report.

Looking at the code, nothing stands out as being a single obvious problem. I use `BLPOP` in a private codebase that uses this library in a similar configuration to this, so that particular command is (in-principal) fine. There must be something else going on here...

Some potential things that might be an issue:

1. Can we be sure the Redis server is still running? The Tokio log-message `dropping I/O source` can be seen when the connection has dropped. The library will attempt to re-connect, but until it has completed the reconnection you'll get those `Redis(EndOfStream)` errors returned. (See also: #30 I'm planning on changing this slightly to be more obvious in a future release.)

2. The connection must have initially worked as the `Redis(Remote("ERR Protocol error: expected \'$\', got \':\'"))` log message is a remote error, the text is produced by the Redis server.

3. It looks like you are using Futures 0.3 with 0.1 compatibility shims? I haven't tested with this configuration, so there's a chance there's an undiagnosed problem there, but I'm not expecting there to be any.

Regarding point 2 - the error is saying Redis received an integer when it was expecting a String. This is probably line 88 in your example, `BLPOP` expects the timeout value to be a String, if you replace `0` with `"0"` that error should go away. Although this shouldn't fix the other problems you've seen, if it does, then there's a bug in how errors are handled.

**Comment by Texlo-Dev** [comment]
Hi Ben,
So I ensured that the Redis server was running, and the compatibility shims are fine. Once I replaced the 0 with a string "0", the error went away!

All in all, it seems like the BLPOP value must be a string, otherwise the Redis server will error. So this does support your theory of a bug in error handling.

**Comment by benashford** [comment]
Cool, thanks for confirming. All errors seem to be dropping the Redis connection rather than just I/O errors.

I'll leave this issue open to track that until it's fixed.

**Comment by benashford** [comment]
The recently released `0.5.0` version now only drops the Redis connection on specific errors rather than all errors. This means that `ERR` messages from the Redis server will keep the connection in-tact.
